### PR TITLE
[experimental] simple-cache-provider

### PR DESCRIPTION
**Pull request by acdlite**
Pushing an early version of this for testing and demonstration purposes.

### CODE REVIEW

File: packages/simple-cache-provider/src/SimpleCacheProvider.js [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+  preload(key: K): void,
+|};
+
+export function createResource<K, V, H>(
+  loadResource: K => V,
+  hash?: K => H = identity,
+): Resource<K, V> {
+  // The read function itself serves as the resource type.
+  function read(cache, key) {
+    invariant(
```
Comment (sebmarkbage) [inline packages/simple-cache-provider/src/SimpleCacheProvider.js]: Slow. This is a suuuuper hot path. Make it `__DEV__` only (and therefore the whole $$typeof can be dev only).

File: packages/simple-cache-provider/src/SimpleCacheProvider.js [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+      isCache(cache),
+      'read(): The first argument must be a cache. Instead received: %s',
+      cache,
+    );
+    if (!isCache(cache)) {
```
Comment (sebmarkbage) [inline packages/simple-cache-provider/src/SimpleCacheProvider.js]: Impossible.

Comment (acdlite) [reply packages/simple-cache-provider/src/SimpleCacheProvider.js]: ?

Comment (acdlite) [reply packages/simple-cache-provider/src/SimpleCacheProvider.js]: Oh, right, idk what these lines are haha. Forgot to delete or something.

Comment (sebmarkbage) [reply packages/simple-cache-provider/src/SimpleCacheProvider.js]: Currently you have the invariant above that guarantees that isCache(cache) is always true.

**Comment by acdlite** [comment]
@sebmarkbage Updated

File: packages/simple-cache-provider/index.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+ */
+
+'use strict';
+
+const SimpleCacheProvider = require('./src/SimpleCacheProvider');
+
+// TODO: decide on the top-level export form.
+// This is hacky but makes it work with both Rollup and Jest.
+module.exports = SimpleCacheProvider.default
+  ? SimpleCacheProvider.default
```
Comment (gaearon) [inline packages/simple-cache-provider/index.js]: Let’s not duplicate this hack anymore. It’s only necessary for legacy stuff.

See Brian’s react-is package for how we should do it in the future.

File: packages/simple-cache-provider/src/SimpleCacheProvider.js [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+        case Empty:
+          // Load the requested resource.
+          throw load(record, key, miss);
+        case Resolved:
+          return ((record.value: any): V);
+        case Rejected:
+        default:
+          // The requested resource previously failed loading.
+          // TODO: When should we try again?
+          throw record.error;
```
Comment (gaearon) [inline packages/simple-cache-provider/src/SimpleCacheProvider.js]: Ideally we should keep the error state and let the user display the message with a “try again” button. Is this what it does?

Comment (acdlite) [reply packages/simple-cache-provider/src/SimpleCacheProvider.js]: Sort of. The error is cached so the next time that resource is read, the cache will throw the error so the user can capture it with an error boundary. But I don’t know when we should reset it to empty. I guess right here, right before we throw? But what if multiple components attempt to read the same value? They should all throw, not just the first one. Maybe we should reset it after n milliseconds.

File: packages/simple-cache-provider/src/SimpleCacheProvider.js [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+}
+
+type ResourceReader<K, V> = (Cache, K) => V;
+
+type Resource<K, V> = ResourceReader<K, V> & {
+  preload(cache: Cache, key: K): void,
+};
+
+export function createResource<K, V, H>(
+  loadResource: K => Promise<V>,
```
Comment (sebmarkbage) [inline packages/simple-cache-provider/src/SimpleCacheProvider.js]: Could add a DEV warning if this is not `typeof loadResource === 'function'`.

File: packages/simple-cache-provider/src/SimpleCacheProvider.js [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+  // $FlowFixMe: Flow complains that K !== H.
+  hash?: K => H = identity,
```
Comment (sebmarkbage) [inline packages/simple-cache-provider/src/SimpleCacheProvider.js]: Do we trust Babel to do the default efficiently?

File: packages/simple-cache-provider/src/SimpleCacheProvider.js [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+
+type ResourceReader<K, V> = (Cache, K) => V;
+
+type Resource<K, V> = ResourceReader<K, V> & {
+  preload(cache: Cache, key: K): void,
+};
+
+export function createResource<K, V, H>(
+  loadResource: K => Promise<V>,
+  // $FlowFixMe: Flow complains that K !== H.
```
Comment (sebmarkbage) [inline packages/simple-cache-provider/src/SimpleCacheProvider.js]: Flow is right.

The return value of the hash should be limited to `string | number | boolean | undefined | null` (i.e. a primitive).

So therefore, if K is not a primitive, it is invalid to not pass a hash function. That's not modeled right now.

Comment (sebmarkbage) [reply packages/simple-cache-provider/src/SimpleCacheProvider.js]: You could possibly model this as overloaded functions.

```js
type primitive = string | number | boolean | void | null;
declare createResource<V>(loadResource: primitive => Promise<V>, hash?: primitive => primitive : V;
declare createResource<V>(loadResource: object => Promise<V>, hash: object => primitive) : V;
declare createResource<V>(loadResource: function => Promise<V>, hash: function => primitive) : V;
```

File: packages/simple-cache-provider/src/SimpleCacheProvider.js (hunk 6 above)
Comment (acdlite) [reply packages/simple-cache-provider/src/SimpleCacheProvider.js]: Ah you're right, Babel uses `arguments`. I'll change.

File: packages/simple-cache-provider/src/SimpleCacheProvider.js (hunk 7 above)
Comment (acdlite) [reply packages/simple-cache-provider/src/SimpleCacheProvider.js]: I should use a sum type for Record, too.

**Comment by acdlite** [comment]
@sebmarkbage Updated

**Comment by acdlite** [comment]
Ready for another look

File: packages/simple-cache-provider/src/SimpleCacheProvider.js [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+      );
+    }
+    if (hash === undefined) {
+      if (__DEV__) {
+        warnIfNonPrimitiveKey(key, 'read');
+      }
+      return cache.read(read, key, loadResource);
+    }
+    const hashedKey = hash(key);
+    return cache.read(read, hashedKey, h => loadResource(key));
```
[the remark this answers is not in the document]
Comment (acdlite) [reply packages/simple-cache-provider/src/SimpleCacheProvider.js]: We could pass both the key and the hash to the cache.

[the remark this answers is not in the document]
Comment (acdlite) [reply packages/simple-cache-provider/src/SimpleCacheProvider.js]: For the simple case, they’re both the same thing. Then we don’t need the closure.

**Comment by acdlite** [comment]
@sebmarkbage Got rid of the closure

**Pull request merged**